### PR TITLE
[CALCITE-4027] Add -Doverwrite option to DiffRepository 

### PR DESCRIPTION
**Pull request by xy2953396112**
https://issues.apache.org/jira/browse/CALCITE-4027

### CODE REVIEW

**Comment by vlsi** [comment]
`overwrite` is a too broad option name.

Please ensure the option works from the command line, and it is documented as well.

File: core/src/test/java/org/apache/calcite/test/DiffRepository.java [hunk 1]
```diff
@@ -762,6 +764,20 @@ public static DiffRepository lookup(Class<?> clazz,
...
+    final String overwrite = System.getProperty("overwrite");
+    // Not need to add test cases manually when you add VM options.
+    // For example: -Doverwrite=true
+    if (overwrite != null && overwrite.equalsIgnoreCase("true")) {
+      try {
+        URL refFile = findFile(clazz, ".xml");
+        final String refFilePath = Sources.of(refFile).file().getAbsolutePath();
+        final String logFilePath = refFilePath.replace(".xml", "_actual.xml");
+        final File logFile = new File(logFilePath);
+        FileUtils.copyFile(logFile, new File(refFilePath.replace("out", "src")));
```
Comment (hsyuan) [inline core/src/test/java/org/apache/calcite/test/DiffRepository.java:776]: Can we use guava `Files.copy`?

**Comment by vlsi** [comment]
@xy2953396112 , please have a look at https://github.com/apache/calcite/blob/978bb7ea44969351468d1b5e240e8f57af7e5770/core/src/main/java/org/apache/calcite/config/CalciteSystemProperty.java to get inspiration for the property name.

File: core/src/test/java/org/apache/calcite/test/DiffRepository.java (hunk 1 above)
Comment (xy2953396112) [reply core/src/test/java/org/apache/calcite/test/DiffRepository.java]: ok

**Comment by xy2953396112** [comment]
@vlsi  Thanks, How about `calcite.test.core.overwrite` ?

**Comment by xy2953396112** [comment]
> @xy2953396112 With your patch, how do I update RelOptRulesTest expected file automatically?
> Say, I added a test in RelOptRulesTest, and specify the property in IDE, it doesn't update the expected file.

Thanks, I'm not sure if `calcite.test.xml.overwrite` is appropriate, maybe a better property name.

**Comment by hsyuan** [comment]
I am not questioning on the name. Can you show me how to update the expected xml file automatically if I added a test case into `RelOptRulesTest`?

**Comment by xy2953396112** [comment]
> I am not questioning on the name. Can you show me how to update the expected xml file automatically if I added a test case into `RelOptRulesTest`?

If we add a test case into `RelOptRulesTest`，we can add a VM options `-Dcalcite.test.xml.overwrite=true` in IDE and run this test case, and we can see the `RelOptRulesTest.xml` has been updated automatically. When we run the test again, it will pass.

**Comment by hsyuan** [comment]
Am I doing it wrong?
There is still error:
```
error while creating xml parser
java.lang.RuntimeException: error while creating xml parser
	at org.apache.calcite.test.DiffRepository.<init>(DiffRepository.java:219)
	at org.apache.calcite.test.DiffRepository.<init>(DiffRepository.java:131)
	at org.apache.calcite.test.DiffRepository$Key.toRepo(DiffRepository.java:833)
	at org.apache.calcite.test.DiffRepository.lookup(DiffRepository.java:770)
	at org.apache.calcite.test.DiffRepository.lookup(DiffRepository.java:734)
	at org.apache.calcite.test.DiffRepository.lookup(DiffRepository.java:720)
	at org.apache.calcite.test.RelOptRulesTest.getDiffRepos(RelOptRulesTest.java:229)
	at org.apache.calcite.test.RelOptTestBase.checkPlanning(RelOptTestBase.java:83)
``` 

Here is the settings:
image

**Comment by hsyuan** [comment]
@xy2953396112 Can you update this pull request according to Julian's comments in the JIRA?

**Comment by xy2953396112** [comment]
> @xy2953396112 Can you update this pull request according to Julian's comments in the JIRA?

OK, I'll update the code later.

**Comment by xy2953396112** [comment]
> Am I doing it wrong?
> There is still error:
> 
> ```
> error while creating xml parser
> java.lang.RuntimeException: error while creating xml parser
> 	at org.apache.calcite.test.DiffRepository.<init>(DiffRepository.java:219)
> 	at org.apache.calcite.test.DiffRepository.<init>(DiffRepository.java:131)
> 	at org.apache.calcite.test.DiffRepository$Key.toRepo(DiffRepository.java:833)
> 	at org.apache.calcite.test.DiffRepository.lookup(DiffRepository.java:770)
> 	at org.apache.calcite.test.DiffRepository.lookup(DiffRepository.java:734)
> 	at org.apache.calcite.test.DiffRepository.lookup(DiffRepository.java:720)
> 	at org.apache.calcite.test.RelOptRulesTest.getDiffRepos(RelOptRulesTest.java:229)
> 	at org.apache.calcite.test.RelOptTestBase.checkPlanning(RelOptTestBase.java:83)
> ```
> 
> Here is the settings:
> image

Add -D `calcite.test.xml.overwrit` option in JUnit and  run a test case in RelOptRulesTest.xml.
image
If we run a test case through Gradle, same as your error info, should fix it.

**Comment by xy2953396112** [comment]
> I made some review comments in https://issues.apache.org/jira/browse/CALCITE-4027. I don't think you've seen them.

Thank you very much for your comment.

**Pull request closed without merging**